homeassistant/components/duwi/base.py

```diff
@@ -0,0 +1,67 @@
+"""Base class for DUWI device types integration."""
+
+from typing import Any
+
+from duwi_smarthome_sdk.device_scene_models import CustomerDevice
+from duwi_smarthome_sdk.manager import Manager
+
+from homeassistant.helpers.device_registry import DeviceInfo
+from homeassistant.helpers.dispatcher import async_dispatcher_connect
+from homeassistant.helpers.entity import Entity
+
+from .const import DOMAIN, DUWI_HA_SIGNAL_UPDATE_ENTITY, MANUFACTURER
+
+
+class DuwiEntity(Entity):
+    """Duwi base device."""
+
+    _attr_has_entity_name = True
+    _attr_should_poll = False
+
+    def __init__(self, device: CustomerDevice, device_manager: Manager) -> None:
+        """Initialize the Duwi entity."""
+        self._attr_unique_id = f"duwi.{device.device_no}"
+        self.device = device
+        self.entity_id = f"{DOMAIN}.{device.device_no}"
+        self.device_manager = device_manager
+        self.is_control = True
+
+    @property
+    def device_info(self) -> DeviceInfo:
+        """Return a device description for the device registry."""
+        return DeviceInfo(
+            identifiers={(DOMAIN, self.device.device_no)},
+            manufacturer=MANUFACTURER,
+            name=f"{self.device.room_name + ' ' if self.device.room_name else 'default room'} {self.device.device_name}",
+            model=self.device.device_type,
+            suggested_area=f"{self.device.floor_name} {self.device.room_name}".strip(),
+        )
+
+    async def async_added_to_hass(self) -> None:
+        """Handle the addition of the entity to hass."""
+        self.async_on_remove(
+            async_dispatcher_connect(
+                self.hass,
+                f"{DUWI_HA_SIGNAL_UPDATE_ENTITY}_{self.device.device_no}",
+                self.handle_signal,
+            )
+        )
+
```

Comment:
When is this called? I am missing the part where the function where the dispatcher is called is registered to the library

Reply:
After enabling WebSocket listening with:

```python
hass.loop.create_task(manager.ws.listen())
```

the device statuses within `manager` are continuously updated to reflect the latest state. This, in turn, triggers the `SharingDeviceListener` to execute the `update_device` method, which sends a notification signal:

```python
dispatcher_send(self.hass, f"{DUWI_HA_SIGNAL_UPDATE_ENTITY}_{device.device_no}")
```

If Home Assistant has a device with this `device_no`, it will listen for this update signal and execute:

```python
self.sync_write_ha_state()
```

This updates the device's state in Home Assistant to reflect the latest changes from the WebSocket feed.